### ISSUE TITLE
m4a metadata create/modified dates for youtube downloaded audio streams recently became invalid

### ISSUE THREAD

**Issue by bughit** [yt-dlp/yt-dlp#9152]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

USA

### Provide a description that is worded well enough to be understood

1. Download a youtube audio stream (e.g. yt-dlp -vUf 599 78Ok52tMB30)
2. examine the timestamps in the m4a metadata using https://exiftool.org/
   Create Date                     : 0000:00:00 00:00:00
   Modify Date                     : 0000:00:00 00:00:00
   Track Create Date               : 0000:00:00 00:00:00
   Track Modify Date               : 0000:00:00 00:00:00
   Media Create Date               : 0000:00:00 00:00:00
   Media Modify Date               : 0000:00:00 00:00:00
3. older downloaded files show valid timestamps

They should either be valid or at least missing if valid ones can't be retrieved. Other software may rely on them. In the mean time, is there an option to force omit them from the m4a?

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vUf', '599', '78Ok52tMB30']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.12.30 from yt-dlp/yt-dlp [f10589e34]
[debug] Lazy loading extractors is disabled
[debug] Python 3.12.1 (CPython AMD64 64bit) - Windows-10-10.0.19045-SP0 (OpenSSL 3.0.11 19 Sep 2023)
[debug] exe versions: ffmpeg 5.1.2-full_build-www.gyan.dev (setts), ffprobe 5.1.2-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.20.0, brotli-1.1.0, certifi-2024.02.02, mutagen-1.47.0, requests-2.31.0, sqlite3-3.43.1, urllib3-2.2.0, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets
[debug] Loaded 1832 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: stable@2023.12.30 from yt-dlp/yt-dlp
yt-dlp is up to date (stable@2023.12.30 from yt-dlp/yt-dlp)
[youtube] Extracting URL: 78Ok52tMB30
[youtube] 78Ok52tMB30: Downloading webpage
[youtube] 78Ok52tMB30: Downloading ios player API JSON
[youtube] 78Ok52tMB30: Downloading android player API JSON
[youtube] 78Ok52tMB30: Downloading m3u8 information
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, vcodec:vp9.2, channels, acodec, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, quality, res, fps, hdr:12(7), source, vcodec:vp9.2(10), channels, acodec, lang, proto, size, br, asr, vext, aext, hasaud, id
[info] 78Ok52tMB30: Downloading 1 format(s): 599
[debug] Invoking http downloader on "<<EXCISED>>"
[debug] File locking is not supported. Proceeding without locking
[download] Destination: @CallMeKevin explores the wild world of simulation games [78Ok52tMB30].m4a
[download] 100% of 1002.64KiB in 00:00:00 at 6.59MiB/s
[FixupM4a] Correcting container of "@CallMeKevin explores the wild world of simulation games [78Ok52tMB30].m4a"
[debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:@CallMeKevin explores the wild world of simulation games [78Ok52tMB30].m4a" -map 0 -dn -ignore_unknown -c copy -f mp4 -movflags "+faststart" "file:@CallMeKevin explores the wild world of simulation games [78Ok52tMB30].temp.m4a"
```

**Comment by bashonly** [comment]
Try `--postprocessor-args "FixupM4a:-map_metadata -1"`

Note that these dates are what youtube sets, not what yt-dlp sets

**Comment by bughit** [comment]
If youtube is sending bad timestamps, yt-dlp should be omitting them from the metadata by default. You could argue this is a feature request, but it's not just a question.

**Comment by bashonly** [comment]
can ffmpeg/ffprobe even output this metadata?

**Comment by bughit** [comment]
> can ffmpeg/ffprobe even output this metadata?

Don't know but I'd guess yes, this is public/standard stuff.

> This seems like either an ffmpeg issue

ffmpeg is a helper tool that does what's instructed, its not it job to autonomously make such decisions.

> something that should be handled by a plugin

Why? You're saying that by default yt-dlp should continue to produce files with garbage metadata. How's that a good default?

**Comment by bughit** [comment]
Please do not mischaracterize this issue. It's not a question, its at least an `enhancement`.

**Comment by bashonly** [comment]
What happens if you add `--fixup never` to your command?

**Comment by bashonly** [comment]
also, please include a FULL verbose log. Your original log has omitted your config options and exe versions

**Comment by bughit** [comment]
> also, please include a FULL verbose log

Updated

**Comment by bughit** [comment]
> What happens if you add `--fixup never` to your command?

The metadata timestamps are correct:

```
Create Date                     : 2024:02:03 10:59:26
Modify Date                     : 2024:02:03 10:59:26
Track Create Date               : 2024:02:03 10:59:26
Track Modify Date               : 2024:02:03 10:59:26
Media Create Date               : 2024:02:03 10:59:26
Media Modify Date               : 2024:02:03 10:59:26
```

So it looks like youtube is still sending good timestamps but it's getting corrupted somehow between yt-dlp and ffmpeg.

I've been using ffmpeg 5.1.2 (encountered errors with 6), but just tried 6.1.1, same zeroed timestamps.

So now that you know the exact yt-dlp and ffmpeg versions and command, can you repro?

**Comment by bughit** [comment]
At this point, its probably no longer an `enhancement`. Youtube is still sending the timestamps. That they are being zeroed by the fixup is either a yt-dlp or an ffmpeg bug. If it turns out ffmpeg, someone from yt-dlp should report it, you will be able to explain it much better.

**Comment by bashonly** [comment]
I dug into this further and here is what I found:

1. If the m4a format is downloaded with `--fixup never`, the resulting file will contain a `creation_time` metadata tag with the correct value:

```shell
$ ffprobe -hide_banner -show_entries format_tags -print_format json unfixed.m4a 2>/dev/null
```
```json
{
    "format": {
        "tags": {
            "major_brand": "dash",
            "minor_version": "0",
            "compatible_brands": "iso6mp41",
            "creation_time": "2024-02-03T10:59:26.000000Z"
        }
    }
}
```

2. if we allow the fixup postprocessor to run on the download, we lose the `creation_time` tag altogther:

```shell
$ ffprobe -hide_banner -show_entries format_tags -print_format json fixed.m4a 2>/dev/null
```
```json
{
    "format": {
        "tags": {
            "major_brand": "isom",
            "minor_version": "512",
            "compatible_brands": "isomiso2mp41",
            "encoder": "Lavf60.16.100"
        }
    }
}
```

3. exiftool will output zeroed-out "Modified at" etc values if the `creation_time` metadata tag is not present in the m4a file. There are no bogus values, they just don't exist, and exiftool compensates by displaying 00:00:00 etc

4. The `FixupM4a` postprocessor is basically the below ffmpeg command line: (note that there are no `-map_metadata` or `-metadata` args present in the command)
```
ffmpeg -y -loglevel repeat+info -i file:input.m4a -map 0 -dn -ignore_unknown -c copy -f mp4 -movflags +faststart file:input.temp.m4a
```

5. The [ffmpeg docs](https://ffmpeg.org/ffmpeg.html) for `-map_metadata` says:
> By default, global metadata is copied from the first input file, per-stream and per-chapter metadata is copied along with streams/chapters. These default mappings are disabled by creating any mapping of the relevant type. 

6. So since we are not passing any metadata mappings in our fixup ffmpeg command line, I think it *should* be copied by default. But that is not happening.

7. If I pass `--postprocessor-args "FixupM4a:-map_metadata 0"` to yt-dlp, it will append `-map_metadata 0` to the ffmpeg args for the `FixupM4a` PP, and we end up with this:

```shell
$ ffprobe -hide_banner -show_entries format_tags -print_format json manual_map_metadata.m4a 2>/dev/null
```
```json
{
    "format": {
        "tags": {
            "major_brand": "isom",
            "minor_version": "512",
            "compatible_brands": "isomiso2mp41",
            "creation_time": "2024-02-03T10:59:26.000000Z",
            "encoder": "Lavf60.16.100"
        }
    }
}
```

So in conclusion:

- I am not sure if this is an ffmpeg bug, or if it is intended behavior since the original `creation_time` value is no longer true/accurate once ffmpeg remuxes it. This is a question for ffmpeg devs, IMO.

- The workaround for the time being is adding `--postprocessor-args "FixupM4a:-map_metadata 0"` to your yt-dlp command.

**Comment by bughit** [comment]
> The workaround for the time being is adding --postprocessor-args "FixupM4a:-map_metadata 0" to your yt-dlp command.

That worked, thanks.

> I think it should be copied by default. But that is not happening.

If I understood you correctly, not only should it be copied per the docs but you agree that's the desired behavior? From the user perspective the fixup is an implementation details so the original creation timestamp should be preserved.

**Comment by bashonly** [comment]
>  not only should it be copied per the docs but you agree that's the desired behavior? 

What I meant is either it's an ffmpeg bug, or it's intended-but-undocumented ffmpeg behavior

> older downloaded files show valid timestamps

This seems worth investigating. Are they youtube downloads? From how long ago? If you download them again, do the new downloads still show valid timestamps? Can you show ffprobe output of one of the old files?

**Comment by bughit** [comment]
> What I meant is either it's an ffmpeg bug, or it's intended-but-undocumented ffmpeg behavior

Ok but regardless of ffmpeg defaults, can we agree that preserving the created_time through fixup should be desired yt-dlp behavior?

An older YT download from 2023-04-28. Don't know which yt-dlp and ffmpeg versions were used:

old metadata:
```
  Metadata:
    major_brand     : M4A
    minor_version   : 0
    compatible_brands: mp42mp41isomiso2
    creation_time   : 2023-04-28T05:51:46.000000Z
  Duration: 00:01:54.71, start: 0.000000, bitrate: 129 kb/s
  Stream #0:0[0x1](und): Audio: aac (LC) (mp4a / 0x6134706D), 44100 Hz, stereo, fltp, 127 kb/s (default)
    Metadata:
      creation_time   : 2019-12-24T02:24:31.000000Z
      vendor_id       : [0][0][0][0]
```

This is probably just chnaging ffmpeg defaults. It appears to have preserved the stream creation_time but generated a new creation_time for the container. This is also not ideal. The timestamps should be the same and inherited from youtube, at least by default.

new re-download metadata:
```
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2mp41
    encoder         : Lavf60.16.100
  Duration: 00:01:54.71, start: 0.000000, bitrate: 129 kb/s
  Stream #0:0[0x1](und): Audio: aac (LC) (mp4a / 0x6134706D), 44100 Hz, stereo, fltp, 127 kb/s (default)
    Metadata:
      handler_name    : ISO Media file produced by Google Inc.
      vendor_id       : [0][0][0][0]
```

**Comment by barsnick** [comment]
> ```
>   Metadata:
>     major_brand     : M4A
>     minor_version   : 0
>     compatible_brands: mp42mp41isomiso2
>     creation_time   : 2023-04-28T05:51:46.000000Z
>   Duration: 00:01:54.71, start: 0.000000, bitrate: 129 kb/s
>   Stream #0:0[0x1](und): Audio: aac (LC) (mp4a / 0x6134706D), 44100 Hz, stereo, fltp, 127 kb/s (default)
>     Metadata:
>       creation_time   : 2019-12-24T02:24:31.000000Z
>       vendor_id       : [0][0][0][0]
> ```
> 
> 
>     
>   
> 
> This is probably just chnaging ffmpeg defaults. It appears to have preserved the stream creation_time but generated a new creation_time for the container. This is also not ideal. The timestamps should be the same and inherited from youtube, at least by default.

If the container of this older download doesn't have an `encoder:` tag, it very likely wasn't remuxed by ffmpeg at all.
